### PR TITLE
Add array support using definitions in the _meta field.

### PR DESCRIPTION
**Pull request by brianolsen87**
Still needs a few more tests and discussion.

Fixes: #2441

### CODE REVIEW

File: presto-elasticsearch/src/test/java/io/prestosql/elasticsearch/TestElasticsearchIntegrationSmokeTest.java [hunk 1]
```diff
@@ -130,6 +130,142 @@ public void testNestedFields()
...
+    public void testArrayFields()
+    {
+        String indexName = "test_arrays";
+
+        embeddedElasticsearchNode.getClient()
+                .admin()
+                .indices()
+                .prepareCreate(indexName)
+                .addMapping("doc", "" +
+                                "{\n" +
```
Comment (martint) [inline presto-elasticsearch/src/test/java/io/prestosql/elasticsearch/TestElasticsearchIntegrationSmokeTest.java]: The `\n` make the document harder to read. I'd get rid of them.

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/client/ElasticsearchClient.java [hunk 2]
```diff
@@ -449,15 +450,17 @@ public IndexMetadata getIndexMetadata(String index)
                     mappings = mappings.elements().next();
                 }
 
-                return new IndexMetadata(parseType(mappings.get("properties")));
+                JsonNode metaNode = nullSafeNode(mappings, "_meta");
+
+                return new IndexMetadata(parseType(mappings.get("properties"), nullSafeNode(metaNode, "presto_meta")));
```
Comment (martint) [inline presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/client/ElasticsearchClient.java]: Let's call this just `presto`. It's understood it's metadata, as it appears in the `_meta` section.

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/ArrayDecoder.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public void decode(SearchHit hit, Supplier<Object> getter, BlockBuilder output)
+    {
+        Object data = getter.get();
+
+        if (data == null) {
+            output.appendNull();
+        }
+        else if (data instanceof List) {
+            final BlockBuilder array = output.beginBlockEntry();
```
Comment (martint) [inline presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/ArrayDecoder.java]: `final` should not be needed here.

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/ArrayDecoder.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            ((List) data).forEach(element -> elementDecoder.decode(hit, () -> element, array));
+            output.closeEntry();
+        }
+        else {
+            throw new PrestoException(TYPE_MISMATCH, "Expected list of primitives for ARRAY field");
```
Comment (martint) [inline presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/ArrayDecoder.java]: Technically, it could be composite objects (e.g., documents), so I'd rephrase as "... list of elements ..."

**Comment by martint** [comment]
BTW, we should add some docs describing what needs to be added to the mapping to declare an array field. Ideally, we could include some examples of calling ES APIs using `curl`.

**Comment by brianolsen87** [comment]
I just pushed those changes. I'll start working on the docs.

**Comment by brianolsen87** [comment]
Okay, added docs. Let me know if I should change the wording/examples

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst [hunk 5]
```diff
@@ -185,6 +185,51 @@ Elasticsearch Presto
 (all others)  (unsupported)
 ============= =============
 
+Array Types
+^^^^^^^^^^^
+
+Elasticsearch has `no dedicated array type <https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html>`_ so the `_meta field <https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-meta-field.html>`_ of the index is used to store an property structure that contains
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst]: ideally wrap paragraphs at 80 char

Comment (mosabua) [reply presto-docs/src/main/sphinx/connector/elasticsearch.rst]: also .. is use to store a property structure

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst [hunk 6]
```diff
@@ -185,6 +185,51 @@ Elasticsearch Presto
...
+the ``isArray`` property to specify which fields in the mapping are to be treated as array types.
+
+Example:
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst]: For example, if you can have an Elasticsearch index that contains documents with the following structure:

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst [hunk 7]
```diff
@@ -185,6 +185,51 @@ Elasticsearch Presto
...
+        "long_field": 314159265359,
+        "id_field": "564e6982-88ee-4498-aa98-df9e3f6b6109",
+        "timestamp_field": "1987-09-17T06:22:48.000Z",
+        "object_field": {
+            "array_field": [86,75,309],
+            "int_field": 2
+        }
+    }
+
+Then you would add the following field property definition to the target index mapping's ``_meta.presto`` property.
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst]: This structure can be defined by using the following command to add the field property definition to the ``_meta.presto`` property of the target index mapping.

**Comment by brianolsen87** [comment]
Thanks @mosabua! I updated the docs locally and going to wait to see if @martint has any more suggestions on the docs before pushing.

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst (hunk 7 above)
Comment (brianolsen87) [reply presto-docs/src/main/sphinx/connector/elasticsearch.rst]: I did change this a little bit by starting with "The array fields of this structure ..." to add a little more clarity.

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst (hunk 5 above)
Comment (martint) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst]: This sentence is a bit long. I'd restructure as follows (or something along those lines): "Fields in Elasticsearch can contain zero or more values, but there is no dedicated array type. To indicate a field contains an array, it can be annotated in a presto-specific structure in the _meta section of the index mapping."

**Comment by brianolsen87** [comment]
Updated the docs and pushed.

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst [hunk 8]
```diff
@@ -185,6 +185,52 @@ Elasticsearch Presto
 (all others)  (unsupported)
 ============= =============
 
+Array Types
+^^^^^^^^^^^
+
+Fields in Elasticsearch can contain `zero or more values <https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html>`_
+, but there is no dedicated array type. To indicate a field contains an array, it can be annotated in a presto-specific structure in
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst]: Presto-specific

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst [hunk 9]
```diff
@@ -185,6 +185,52 @@ Elasticsearch Presto
...
+the `_meta <https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-meta-field.html>`_ section of the index mapping.
+
+For example, if you can have an Elasticsearch index that contains documents with the following structure:
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst]: For example, you can have an ...

Comment (mosabua) [reply presto-docs/src/main/sphinx/connector/elasticsearch.rst]: otherwise the current one is not a full sentence

**Comment by martint** [comment]
Great job! Thanks for implementing this. I merged it as e5bd9ebebb6034ae56eced2abc0611e05d38fc16

**Comment by brianolsen87** [comment]
Awesome! Thanks @martint and @mosabua!

**Comment by mosabua** [comment]
Nice work!
